src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,467 @@
+
+                    for channel_id in expiring_channels {
+                        channels.remove(&channel_id);
+                    }
+                }
+            });
+        }
+
+        Self {
+            max_pending_acks: u64::from(config.max_pending_acks),
+            max_pending_acks_per_channel: u64::from(config.max_pending_acks_per_channel),
+            max_number_of_ack_channels: u64::from(config.max_number_of_ack_channels),
+            channels,
+            shutdown,
+            total_pending_acks: AtomicU64::new(0),
+        }
+    }
+
+    /// Creates a channel with the specified id if it does not exist.
+    async fn create_or_get_channel(&self, id: String) -> Result<Arc<Channel>, Rejection> {
+        let channels = self.channels.read().await;
+        if let Some(channel) = channels.get(&id) {
+            return Ok(Arc::clone(channel));
+        }
+        drop(channels);
+
+        let mut channels = self.channels.write().await;
+        if channels.len() < self.max_number_of_ack_channels as usize {
+            // Create the channel if it does not exist
+            let channel = Arc::new(Channel::new(
+                self.max_pending_acks_per_channel,
+                self.shutdown.clone(),
+            ));
+            channels.insert(id, Arc::clone(&channel));
+            Ok(channel)
+        } else {
+            Err(Rejection::from(ApiError::ServiceUnavailable))
+        }
+    }
+
+    /// Gets the next available ack id from a specified channel, creating the channel if it does not exist
+    pub async fn get_ack_id_from_channel(
+        &self,
+        channel_id: String,
+        batch_rx: BatchStatusReceiver,
+    ) -> Result<u64, Rejection> {
+        let channel = self.create_or_get_channel(channel_id).await?;
+        let ack_id = channel.get_ack_id(batch_rx);
+        let total_pending_acks = self.total_pending_acks.fetch_add(1, Ordering::Relaxed) + 1;
+        if total_pending_acks > self.max_pending_acks {
+            self.total_pending_acks.fetch_sub(
+                self.drop_oldest_pending_ack_per_channel().await,
+                Ordering::Relaxed,
+            );
+        }
+        Ok(ack_id)
+    }
+
+    /// Gets the requested ack id statuses from a specified channel, creating the channel if it does not exist
+    pub async fn get_acks_status_from_channel(
+        &self,
+        channel_id: String,
+        ack_ids: &[u64],
+    ) -> Result<HashMap<u64, bool>, Rejection> {
+        let channel = self.create_or_get_channel(channel_id).await?;
+        let acks_status = channel.get_acks_status(ack_ids);
+        let dropped_ack_count = acks_status.values().filter(|status| **status).count();
+        self.total_pending_acks
+            .fetch_sub(dropped_ack_count as u64, Ordering::Relaxed);
+        Ok(acks_status)
+    }
+
+    /// Drops the oldest ack id (if one exists) from each channel
+    async fn drop_oldest_pending_ack_per_channel(&self) -> u64 {
```

Comment:
I didn't see in the other code, will this be called in a loop? If so, it would be better to have a parameter of the number of pending acks to drop in order to amortize the work (locking, etc).

Reply:
It's not currently called in a loop. Now that I think about it, it might make sense to call it in a loop until it returns > 0 to ensure acks are actually dropped from `ack_ids_status`. I can see the scenario: 
- client sends requests up to `max_pending_acks` limit 
- requests are waiting to be finalized (`ack_ids_status` is empty) 
- client sends another request which exceeds `max_pending_acks` but `ack_ids_status` is still empty, so nothing is dropped

I think we can parameterize the number of acks to drop in addition to looping. Does that make sense?